### ISSUE TITLE
v1.1.0 and v1.1.1 firmwares broken by bad CLOCK_SOURCE default in atmelavr 3.4.0

### ISSUE THREAD

**Issue by msparks** [jaytavares/cinesense#6]
**TL;DR**: [platformio/platform-atmelavr](https://github.com/platformio/platform-atmelavr) 3.4.0 contains a bug that causes the ATtiny85 to run at 4 MHz instead of 8 MHz, which is enough to completely break Cinesense. The **firmware in the v1.1.1 release is DOA** for me – no auto-off/on behavior at all – but the firmware binary in the v1.0.0 release *does* work. The v1.1.0 and v1.1.1 releases were built (by GitHub Actions) [using the broken atmelavr 3.4.0](https://github.com/jaytavares/cinesense/runs/6116544230?check_suite_focus=true). atmelavr 4.0.0 contains a fix for the bug.

This issue affects builds of Cinesense that used [platformio/platform-atmelavr](https://github.com/platformio/platform-atmelavr) 3.4.0 only. The root cause was [fixed upstream in 4.0.0](https://github.com/platformio/platform-atmelavr/releases/tag/v4.0.0). What follows below is a detailed version of https://github.com/platformio/platform-atmelavr/issues/264 as it applies to Cinesense. (I wrote most of the following before finding that issue. I'm leaving all of the details here in case someone falls into the same bad-package trap that I did.)

### Background

I ran into this problem when I tried flashing the v1.1.1 firmware to my Dec 2021 Cinesense, hoping to pull in the fix for #1. My journey was as follows:

1. Flashed `firmware.hex` from v1.1.1 using `avrdude`. **Result**: Cinesense functionality didn't seem to work at all. Speakers (and interface module LED) would stay on without a signal, and stay off with a signal.
2. Flashed `firmware.hex` from v1.0.0. **Result**: Cinesense worked as expected.
3. Cloned the repo, and ran `pio run` to compile and flash. **Result**: Same as v1.1.1 official release, non-functional.
4. `git checkout v1.0.0`, then `pio run` to run the v1.0.0 code. **Result**: Still broken (!).

After a few other dead ends, I read through the [Theory of Operation](https://github.com/jaytavares/cinesense#theory-of-operation-for-nerds) and then the code. To see if I could get _anything_ working from source, I added this to the top of `loop()` to make the interface module's LED blink:

```c++
pinMode(CONTROL_OUT, OUTPUT);
digitalWrite(CONTROL_OUT, HIGH);
delay(5000);
pinMode(CONTROL_OUT, OUTPUT);
digitalWrite(CONTROL_OUT, LOW);
delay(500);
return;
```

Surprisingly, the LED stayed on for *10* seconds, not *5* seconds as written. That led to discovering https://community.platformio.org/t/board-build-f-cpu-parameter-has-no-effect-on-attiny85/25485, which precisely describes the problem.

### Problem statement

Cinesense sets the ATtiny85 low fuse bits to [`0xE2`](http://eleccelerator.com/fusecalc/fusecalc.php?chip=attiny85&LOW=E2&HIGH=DF&EXTENDED=FF&LOCKBIT=FF) to use the internal oscillator at 8MHz. However, the value of the [`CLOCK_SOURCE` define](https://github.com/SpenceKonde/ATTinyCore#determining-clock-speed-and-source-from-within-the-sketch) can affect the clock prescalar, and in [atmelavr 3.4.0](https://github.com/platformio/platform-atmelavr/releases/tag/v3.4.0), the default for `CLOCK_SOURCE` for ATtiny85 changed [from `0` to `6`](https://github.com/platformio/platform-atmelavr/commit/012de336af56472e65e2db2e6c2d95d9a137803d), where:

> - 0 - Internal 8MHz oscillator, not prescaled, or prescaled to 1 MHz (ie, fully set by fuses)
> - 6 - Internal PLL (x5 and x61 only)

When `CLOCK_SOURCE` is `6`, the atmelavr code believes the 16MHz PLL is being used, and `F_CPU` is set to 8 MHz, so [it sets a clock division of 2 to match 8 MHz](https://github.com/SpenceKonde/ATTinyCore/blob/master/avr/cores/tiny/wiring.c#L1041-L1050) from a supposed 16 MHz clock source. But since the fuses select the internal RC oscillator running at 8 MHz, the division by 2 causes the ATtiny to run at 4 MHz, thus `delay()`s run for twice as long as they should.

### Evidence

Here's what I see when compile Cinesense at HEAD, using PlatformIO with the Atmel AVR 3.4.0 platform:

```
% pio run -t clean
% pio debug -v
Processing attiny85 (platform: atmelavr; board: attiny85; framework: arduino; board_fuses.lfuse: 0xE2; board_fuses.hfuse: 0xDF; board_fuses.efuse: 0xFF; targets: fuses, upload)
------------------------------------------------------------------------------------------------------------------------------------------------------------------------
CONFIGURATION: https://docs.platformio.org/page/boards/atmelavr/attiny85.html
PLATFORM: Atmel AVR (3.4.0) > Generic ATtiny85
HARDWARE: ATTINY85 8MHz, 512B RAM, 8KB Flash
DEBUG: Current (simavr) On-board (simavr)
PACKAGES:
 - framework-arduino-avr-attiny @ 1.5.2
 - toolchain-atmelavr @ 1.70300.191015 (7.3.0)
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain, Compatibility ~ soft
Found 9 compatible libraries
Scanning dependencies...
No dependencies
Building in debug mode
avr-g++ -o .pio/build/attiny85/src/AudioSense.cpp.o -c -fno-exceptions -fno-threadsafe-statics -fpermissive -std=gnu++11 -Wall -ffunction-sections -fdata-sections -flto -mmcu=attiny85 -Og -g2 -ggdb2 -DPLATFORMIO=60002 -DARDUINO_AVR_ATTINYX5 -DNEOPIXELPORT=PORTB -DCLOCK_SOURCE=6 -DF_CPU=8000000L -DARDUINO_ARCH_AVR -DARDUINO=10808 -D__PLATFORMIO_BUILD_DEBUG__ -Iinclude -Isrc -I/Users/ms/.platformio/packages/framework-arduino-avr-attiny/cores/tiny -I/Users/ms/.platformio/packages/framework-arduino-avr-attiny/variants/tinyX5 src/AudioSense.cpp
...
```

Note:
- `board_fuses.lfuse: 0xE2`
- `PLATFORM: Atmel AVR (3.4.0) > Generic ATtiny85`
- `-DCLOCK_SOURCE=6 -DF_CPU=8000000L`

### Solution

**Update `atmelavr` to 4.0.0** to restore the behavior prior to 3.4.0. 4.0.0 reverts the `CLOCK_SOURCE` default change from 3.4.0 (https://github.com/platformio/platform-atmelavr/issues/264).

From within the Cinesense repo:

```
% pio pkg update
Resolving attiny85 environment packages...
Platform Manager: Updating atmelavr @ 3.4.0
Platform Manager: Removing atmelavr @ 3.4.0
Platform Manager: atmelavr@3.4.0 has been removed!
Platform Manager: Installing platformio/atmelavr @ 4.0.0
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Platform Manager: atmelavr@4.0.0 has been installed!
```

(On 3.4.0, the workaround is to add [`build_unflags`](https://docs.platformio.org/en/stable/projectconf/section_env_build.html#build-unflags) to the `env:attiny85` section of `platformio.ini`:

```
build_unflags = -DCLOCK_SOURCE=6
```

This is a temporary workaround though; the stable interface is `build.clock_source`, introduced in 4.0.0.)

After updating:

```
% pio debug -v
Processing attiny85 (platform: atmelavr; board: attiny85; framework: arduino; board_fuses.lfuse: 0xE2; board_fuses.hfuse: 0xDF; board_fuses.efuse: 0xFF; targets: fuses, upload)
...
PLATFORM: Atmel AVR (4.0.0) > Generic ATtiny85
...
avr-g++ -o [...] -DCLOCK_SOURCE=0 [...] src/AudioSense.cpp
...
```

With `CLOCK_SOURCE` fixed, the blink example above had correct timing. I then built and flashed the code at HEAD and that worked as well. (And now I finally have a fix for #1!)

### Conclusion

The `atmelavr` platform at version 3.4.0 produces firmwares at that run Cinesense at 4 MHz. Version 4.0.0 (or a version older than 3.4.0) is required to produce a firmware that runs at the correct frequency (8 MHz).

The `firmware.hex` files in Cinesense v1.1.0 and v1.1.1 were built in 2022-04 by GitHub Actions, and the current version at that time was atmelavr 3.4.0. The [Action logs](https://github.com/jaytavares/cinesense/runs/6116544230?check_suite_focus=true) for the `Run PlatformIO` clearly show the use of the bad version:

```
Run pio ci --board=attiny85 --lib=include --project-option=platformio.ini --keep-build-dir --build-dir=build src
...
Platform Manager: Installing atmelavr
Downloading...
Unpacking...
Platform Manager: atmelavr @ 3.4.0 has been installed!
```

atmelavr 4.0.0, which contains the fix, wasn't released until [2022-05-31](https://github.com/platformio/platform-atmelavr/releases/tag/v4.0.0). Thus, the published firmwares of Cinesense v1.1.0 and v1.1.1 are nonfunctional.

No changes are required to the Cinesense code. And presumably by now the default version of atmelavr 4.0.0, so the next triggered release action should produce a working firmware.

**Comment by jaytavares** [comment]
Thanks for the thorough troubleshooting @msparks. I'm sorry about those broken firmware releases. I've deleted them. To fix going forward I've updated the CI script to peg platformio at the latest stable version (6.1.0). That should prevent future upstream problems from creeping in to our firmware. Would you mind testing the latest v1.1.2 firmware to ensure the problem has been fixed? 

Thanks again for all your sleuth work. 🥇

**Comment by msparks** [comment]
> Would you mind testing the [latest v1.1.2 firmware](https://github.com/jaytavares/cinesense/releases/tag/v1.1.2) to ensure the problem has been fixed?

I just flashed the v1.1.2 firmware and can confirm it behaves as expected! 🎉

The [Actions build of v1.1.2](https://github.com/jaytavares/cinesense/runs/7221259865?check_suite_focus=true) also clearly shows the installation of atmelavr 4.0.0, so the latest release with the fix is being picked up correctly.

> To fix going forward I've updated the CI script to peg platformio at the current stable version (6.1.0).

Note that this pin won't apply to the `atmelavr` platform, only to PlatformIO itself. PlatformIO has its own package manager; during builds, it'll fetch the packages necessary to build for the platforms and frameworks specified in the `platformio.ini` file in the project. To guarantee stability, I think you'd also have to pin atmelavr to a specific version, e.g.,

```ini
[env:attiny85]
platform = atmelavr@4.0.0
...
```

Where `@4.0.0` means *exactly* 4.0.0. It's possible to use something like `@^4.0.0`, which means any `4.x.x` (backwards-compatible) version is acceptable, but of course that wouldn't prevent bugs from being introduced in a `4.x.x` release. (Ref: [Version Requirements](https://docs.platformio.org/en/latest/core/userguide/pkg/cmd_install.html#version-requirements))

In practice, this kind of showstopper bug should be rare (hopefully), and without a forcing function to revisit and update the pinned versions, it's possible that pinning creates more problems than it solves in the long run. So, a reasonable compromise might be to just avoid the known-bad package:

```ini
[env:attiny85]
platform = atmelavr@!=3.4.0
...
```

All that said, thanks for fixing the releases and for creating Cinesense. This was a fun learning experience. :)

**Comment by jaytavares** [comment]
Ah, gotcha. I've pinned it to v4. I'd rather shoot for deterministic builds since this firmware won't be seeing a ton of new development/features. The only real thing left is analog audio sensing which I'm working on currently. 

Thanks again for your work on this.

**Comment by Temtaime** [comment]
So how one should select 16MHz for attiny85 ?
I still get error `framework-arduino-avr-attiny\cores\tiny\wiring.c:1095:8: error: #error "Frequency requested from internal oscillator that cannot be generated by prescaling"` with the latest version, but related bugreport is closed
